File: src/display.rs
use crate::models::Workout;

/// Format meters with comma separators: 1000000 → "1,000,000"
pub fn format_meters(m: u64) -> String {
    let s = m.to_string();
    let mut result = String::new();
    for (i, c) in s.chars().rev().enumerate() {
        if i > 0 && i % 3 == 0 {
            result.push(',');
        }
        result.push(c);
    }
    result.chars().rev().collect()
}

/// Format a percentage with one decimal: 0.1234 → "12.3%"
pub fn format_percent(ratio: f64) -> String {
    format!("{:.1}%", ratio * 100.0)
}

/// Format a workout as a compact log line.
/// Example: "03/02  5500m  31:52.2  2:53.8/500m  26spm  118bpm  83df"
pub fn format_workout_line(w: &Workout, date_format: &str) -> String {
    let date_str = match w.parsed_date() {
        Some(dt) => dt.format(date_format).to_string(),
        None => w.date[..10].to_string(),
    };

    let distance = format!("{}m", format_meters(w.distance));
    let time = &w.time_formatted;
    let pace = w.pace_500m.as_deref().unwrap_or("-");
    let spm = w
        .stroke_rate
        .map(|s| format!("{}spm", s))
        .unwrap_or_else(|| "-".to_string());
    let hr = w
        .heart_rate
        .as_ref()
        .and_then(|h| h.average)
        .map(|h| format!("{}bpm", h))
        .unwrap_or_else(|| "-".to_string());
    let df = w
        .drag_factor
        .map(|d| format!("{}df", d))
        .unwrap_or_else(|| "-".to_string());

    format!(
        "{}  {}  {}  {}/500m  {}  {}  {}",
        date_str, distance, time, pace, spm, hr, df
    )
}

/// Format meters per week for display: 19231 → "19,231m/week"
pub fn format_meters_per_week(m: u64) -> String {
    format!("{}m/week", format_meters(m))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_meters() {
        assert_eq!(format_meters(0), "0");
        assert_eq!(format_meters(500), "500");
        assert_eq!(format_meters(1000), "1,000");
        assert_eq!(format_meters(1_000_000), "1,000,000");
    }

    #[test]
    fn test_format_percent() {
        assert_eq!(format_percent(0.0), "0.0%");
        assert_eq!(format_percent(0.5), "50.0%");
        assert_eq!(format_percent(1.0), "100.0%");
    }
}

File: src/display.ts
import type { Workout } from "./models.ts";
import { pace500m, parsedDate } from "./models.ts";

export function formatMeters(m: number): string {
  return m.toLocaleString("en-US");
}

export function formatPercent(ratio: number): string {
  return `${(ratio * 100).toFixed(1)}%`;
}

export function formatMetersPerWeek(m: number): string {
  return `${formatMeters(m)}m/week`;
}

export function formatDate(d: Date, fmt: string): string {
  // Go-style reference time: "01/02" means month/day
  if (fmt === "01/02" || fmt === "%m/%d") {
    const mm = String(d.getMonth() + 1).padStart(2, "0");
    const dd = String(d.getDate()).padStart(2, "0");
    return `${mm}/${dd}`;
  }
  if (fmt === "%Y-%m-%d") {
    const y = d.getFullYear();
    const mm = String(d.getMonth() + 1).padStart(2, "0");
    const dd = String(d.getDate()).padStart(2, "0");
    return `${y}-${mm}-${dd}`;
  }
  // Fallback: MM/DD
  const mm = String(d.getMonth() + 1).padStart(2, "0");
  const dd = String(d.getDate()).padStart(2, "0");
  return `${mm}/${dd}`;
}

export function formatWorkoutLine(w: Workout, dateFormat: string): string {
  const d = parsedDate(w);
  const dateStr = formatDate(d, dateFormat);
  const distance = `${formatMeters(w.distance)}m`;
  const pace = pace500m(w);
  const spm = w.stroke_rate ? `${w.stroke_rate}spm` : "-";
  const hr = w.heart_rate?.average && w.heart_rate.average > 0 ? `${w.heart_rate.average}bpm` : "-";
  const df = w.drag_factor ? `${w.drag_factor}df` : "-";

  return `${dateStr}  ${distance.padStart(7)}  ${w.time_formatted.padStart(8)}  ${pace.padStart(7)}/500m  ${spm.padStart(5)}  ${hr.padStart(6)}  ${df.padStart(4)}`;
}

export function sparkBar(value: number, max: number): string {
  if (max === 0) return "";
  const BAR_WIDTH = 20;
  const filled = Math.round((value / max) * BAR_WIDTH);
  return "\u2588".repeat(filled) + "\u2591".repeat(BAR_WIDTH - filled);
}

const TREND_THRESHOLD = 0.02;

export function trendArrow(prev: number, curr: number): string {
  if (prev === 0) return " ";
  const diff = (curr - prev) / prev;
  if (diff > TREND_THRESHOLD) return "\u2191";
  if (diff < -TREND_THRESHOLD) return "\u2193";
  return "\u2192";
}

export function paceArrow(prev: number, curr: number): string {
  if (prev === 0) return " ";
  return trendArrow(curr, prev); // reversed: lower pace is better
}
